crates/moonbuild/src/entry.rs

```diff
@@ -92,14 +104,74 @@ fn render_result(result: Option<usize>, quiet: bool, mode: &str) -> anyhow::Resu
     }
 }
 
+#[derive(Default)]
+struct ResultCatcher {
+    content_writer: Vec<String>, // todo: might be better to directly write to string
+    n_warnings: usize,
+    n_errors: usize,
+}
+
+impl ResultCatcher {
+    fn append_content(&mut self, s: impl Into<String>) {
+        self.content_writer.push(s.into());
+    }
+
+    fn new_warn(&mut self) {
+        self.n_warnings += 1;
+    }
+
+    fn new_error(&mut self) {
+        self.n_errors += 1;
+    }
+}
+
+#[allow(clippy::too_many_arguments)] // This is inefficient and we know it
+fn render_and_catch_callback(
+    catcher: Arc<Mutex<ResultCatcher>>,
+    output_json: bool,
+    use_fancy: bool,
+    check_patch_file: Option<PathBuf>,
+    explain: bool,
+    render_no_loc: DiagnosticLevel,
+    source_dir: PathBuf,
+    target_dir: PathBuf,
+) -> impl Fn(&str) {
+    move |output: &str| {
+        output
+            .split('\n')
+            .filter(|it| !it.is_empty())
+            .for_each(|content| {
+                let report_kind = if output_json {
+                    println!("{content}");
+                    None
+                } else {
+                    moonutil::render::MooncDiagnostic::render(
+                        content,
+                        use_fancy,
+                        check_patch_file.clone(),
+                        explain,
+                        render_no_loc,
+                        (source_dir.clone(), target_dir.clone()),
+                    )
+                };
+                let mut catcher = catcher.lock().unwrap();
+                catcher.append_content(content);
+                match report_kind {
```

Comment:
Why are we separating these calls instead of just using `catcher.append_content(report_kind, content)`? Are they intended to be used separately?

Reply:
You are absolutely right (lol)!  Let me fix this ;)